src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -79,30 +462,99 @@ core::smart_refctd_ptr<ICPUPipelineLayout> CGraphicsPipelineLoaderMTL::makePipel
-                blendParams.blendParams[i].srcAlphaFactor = EBF_SRC_ALPHA;
-                blendParams.blendParams[i].dstColorFactor = EBF_ONE_MINUS_SRC_ALPHA;
-                blendParams.blendParams[i].dstAlphaFactor = EBF_ONE_MINUS_SRC_ALPHA;
+                blendParams.blendParams[k].blendEnable = true;
+                blendParams.blendParams[k].srcColorFactor = EBF_SRC_ALPHA;
+                blendParams.blendParams[k].srcAlphaFactor = EBF_SRC_ALPHA;
+                blendParams.blendParams[k].dstColorFactor = EBF_ONE_MINUS_SRC_ALPHA;
+                blendParams.blendParams[k].dstAlphaFactor = EBF_ONE_MINUS_SRC_ALPHA;
+            }
+        }
+
+        const uint32_t j = i*PIPELINE_PERMUTATION_COUNT;
+
+        vtxParams.enabledAttribFlags = (1u << POSITION) | (1u << NORMAL);
+        vtxParams.enabledBindingFlags = 1u << BND_NUM;
+        vtxParams.bindings[BND_NUM].stride = 24u;
+        vtxParams.bindings[BND_NUM].inputRate = EVIR_PER_VERTEX;
+        //position
+        vtxParams.attributes[POSITION].binding = BND_NUM;
+        vtxParams.attributes[POSITION].format = EF_R32G32B32_SFLOAT;
+        vtxParams.attributes[POSITION].relativeOffset = 0u;
+        //normal
+        vtxParams.attributes[NORMAL].binding = BND_NUM;
+        vtxParams.attributes[NORMAL].format = EF_A2B10G10R10_SNORM_PACK32;
+        vtxParams.attributes[NORMAL].relativeOffset = 20u;
+
+        auto layout = makePipelineLayoutFromMtl(materials[i]);
+        auto shaders = getShaders(false);
+        core::smart_refctd_ptr<ICPUDescriptorSet> ds3;
+        {
+            const std::string dsCacheKey = std::string(fullName.c_str())+"?"+materials[i].name+"?_ds";
+            const asset::IAsset::E_TYPE types[]{ asset::IAsset::ET_DESCRIPTOR_SET, (asset::IAsset::E_TYPE)0u };
+            auto ds_bundle = _override->findCachedAsset(dsCacheKey, types, ctx.inner, _hierarchyLevel + ICPUMesh::DESC_SET_HIERARCHYLEVELS_BELOW);
+            if (!ds_bundle.isEmpty())
+                ds3 = core::smart_refctd_ptr_static_cast<ICPUDescriptorSet>(ds_bundle.getContents().first[0]);
+            else 
+            {
+                auto views = loadImages(relPath.c_str(), materials[i], ctx);
+                ds3 = makeDescSet(std::move(views), layout->getDescriptorSetLayout(3u));
+                if (ds3)
+                {
+                    SAssetBundle bundle{ds3};
+                    _override->insertAssetIntoCache(bundle, dsCacheKey, ctx.inner, _hierarchyLevel + ICPURenderpassIndependentPipeline::DESC_SET_HIERARCHYLEVELS_BELOW);
+                }
             }
         }
 
-        m_assetMgr->setAssetMetadata(pipelines[i].get(), core::make_smart_refctd_ptr<CMTLPipelineMetadata>(std::move(materials[i])));
+        pipelines[j] = core::make_smart_refctd_ptr<ICPURenderpassIndependentPipeline>(nullptr, core::smart_refctd_ptr(layout), nullptr, nullptr, vtxParams, blendParams, primParams, rasterParams);
```

Comment:
you should be caching these, as there is only so many permutations you can generate (plus they will be cloned and patched for vertex inputs anyway).

Reply:
i think u need to read the code again

Reply:
they're cached because they're retval of the loader. I can cache any more (more than preventing from `clone()` in OBJ loader) because each mtl file results in unique metadata

Reply:
ah right, unique metadata :(